test/e2e/declarative_test_fwk/helper.py

```diff
@@ -201,16 +235,50 @@ def assert_items(self, expectations: dict, state: dict) -> None:
         if not expectations:
             # nothing to assert as there are no expectations
             return
-        if not state:
-            # there are expectations but no given state to validate
+        if not actual:
+            # there are expectations but no given actual state to validate
             # following assert will fail and assert introspection will provide useful information for debugging
-            assert expectations == state
+            assert expectations == actual
 
         for (key, value) in expectations.items():
             # conditions are processed separately
             if key == "conditions":
                 continue
-            assert (key, value) == (key, state.get(key))
+            # tags are processed by custom logic
+            if key == "tags":
+                for expectedTag in value:
```

Comment:
Can this custom logic be implemented in a dedicated helper function that can be independently tested/validated?

Reply:
Do you want me to add unit tests for those custom logic? There are 2 custom logic for validating k8s resource and 3 custom logic for validating aws resource.